### ISSUE TITLE
How to use Multiple excludes with multiple backup directories 

### ISSUE THREAD

**Issue by No15pecial** [borgbackup/borg#4874]
I have checked the FAQ & googled this. 
The Docs show 

borg create --verbose --filter AME --list --stats --show-rc --compression lz4 --exclude-caches --exclude '/home/*/.cache/*' --exclude '/var/cache/*' --exclude '/var/tmp/*' ::'{hostname}-{now}'  /etc /home  /root /var 

Seems like your declare all the excluded first and then the directories to be backup ud.

I issued the following in my script 

borg create --verbose  --stats --chunker-params 19,23,21,4095 --show-rc --exclude-caches --exclude '/NewCorpFiler-Mounts/ABC_IT/.snapshot'  --exclude '/NewCorpFiler-Mounts/ABC_ENG/.snapshot'  --exclude '/NewCorpFiler-Mounts/Vol1/.snapshot'  --exclude '/NewCorpFiler-Mounts/ABC_ENG_VM-BKS/.snapshot'   /Synology-Mounts/BorgHoard/ABCBorgRepo::BorgBK01-2019-12-12 /NewCorpFiler-Mounts/ABC_IT/ /NewCorpFiler-Mounts/ABC_ENG/ /NewCorpFiler-Mounts/Vol1/ /NewCorpFiler-Mounts/ABC_ENG_VM-BKS/ 

This backup usually run 10ish hours after 15 hours I used htop to see what was going on. I saw that the snapshot directories (specifically /NewCorpFiler-Mounts/ABC_IT/.snapshot/hourly.25) were in fact being accessed  

some of my snapshots occur hourly so if this backup runs for 11 hours and the snapshots changing every hour I'm sure it wreaking havoc. 

When I use one exclude statement --exclude '*/.snapshot'  snapshots are excluded. - I;m sure I'm missing something but I just can't see it. 

## Have you checked borgbackup docs, FAQ, and open Github issues?

yes, even did a google search https://github.com/borgbackup/borg/issues/4053

## Is this a BUG / ISSUE report or a QUESTION?

Invalid

#### Your borg version (borg -V).
borg 1.1.5

#### Operating system (distribution) and version.
Linux  4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux    Ubuntu 18
#### Hardware / network configuration, and filesystems used.
HP Laptop, Net Appliance Filer Synology NAS 

#### How much data is handled by borg?
about 3TB

#### Full borg commandline that lead to the problem (leave away excludes and passwords)
borg create --verbose  --stats --chunker-params 19,23,21,4095 --show-rc --exclude-caches --exclude '/NewCorpFiler-Mounts/ABC_IT/.snapshot'  --exclude '/NewCorpFiler-Mounts/ABC_ENG/.snapshot'  --exclude '/NewCorpFiler-Mounts/Vol1/.snapshot'  --exclude '/NewCorpFiler-Mounts/ABC_ENG_VM-BKS/.snapshot'   /Synology-Mounts/BorgHoard/ABCBorgRepo::BorgBK01-2019-12-12 /NewCorpFiler-Mounts/ABC_IT/ /NewCorpFiler-Mounts/ABC_ENG/ /NewCorpFiler-Mounts/Vol1/ /NewCorpFiler-Mounts/ABC_ENG_VM-BKS/ 

## Describe the problem you're observing.
--excludes not being excluded

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

I believe so as it happened for a few days

**Comment by fantasya-pbem** [comment]
First you should try the same with borg 1.1.10.
You could also use one exclude: ``--exclude /NewCorpFiler-Mounts/*/.snapshot``, if that is applicable.

**Comment by No15pecial** [comment]
"You could also use one exclude: --exclude /NewCorpFiler-Mounts/*/.snapshot, if that is applicable."
My plan is to actually use a nightly or weekly snapshot in the backup so if I wildcard exclude snapshots that won't work.  

So I will be targeting  /NewCorpFiler-Mounts/ABC_IT/  /NewCorpFiler-Mounts/Vol1/ /NewCorpFiler-Mounts/ABC_ENG_VM-BKS/snapshot/Weekly.1/

**Comment by No15pecial** [comment]
Should I be building my command as 

--exclude '/NewCorpFiler-Mounts/ABC_ENG_VM-BKS/.snapshot' /NewCorpFiler-Mounts/ABC_ENG_VM-BKS/  --exclude '/NewCorpFiler-Mounts/ABC_ENG/.snapshot' /NewCorpFiler-Mounts/ABC_ENG/  . . .

**Comment by ThomasWaldmann** [comment]
Just as a general comment: how about if you would try and discuss this based on shorted pathes?

Your pathes are really TL;DR and you could discuss the fundamental problem also using just /a/b/c.

**Comment by No15pecial** [comment]
TL;DR ???   Maybe this would be a quicker read 

I'm looking to
borg create --exclude /a/.snapshot --exclude /b/.snapshot    /a  /b  /c/.snapshot.nightl.1   /d/.snapshot/weekly.1  

What I found was that it did not exclude any of the /a/.snapshot directories   I killed it after 15 hours of running.

**Comment by ThomasWaldmann** [comment]
```
$ borg create --list --exclude /a/.snapshot --exclude /b/.snapshot repo::archive /a /b /c/.snapshot /d/.snapshot
x /a/.snapshot
d /a
x /b/.snapshot
d /b
A /c/.snapshot/file
d /c/.snapshot
A /d/.snapshot/file
d /d/.snapshot

$ borg -V
borg 1.1.5
```

So, works for me.

**Comment by No15pecial** [comment]
Yeah that's great but I did not have the same experience.  I can try it again but it takes a long long time. I have a lot of snapshots. I know you like brevity but I can't test with /a /b /c ... They don't exist on my nas and it would take time for it to build the snapshots if I crate them ... 

I'll have to look for one of my smaller structures to see if I can experiment and if i have to kill it I guess I can list the contents of the specific repo ...  

root@BorgBK01:~# ll /NewCorpFiler-Mounts/ABC_ENG/.snapshot
total 348
drwxrwxrwx 86 root root 8192 Dec 14 15:00 ./
drwxrwxrwx  5 root root 4096 May 21  2018 ../
drwxrwxrwx  4 root root 4096 May 21  2018 2019-12-11-Manual-Snapshot-GDC_ENG/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.0/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.1/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.10/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.11/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.12/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.13/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.14/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.15/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.16/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.17/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.18/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.19/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.2/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.20/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.21/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.22/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.23/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.24/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.25/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.26/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.27/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.3/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.4/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.5/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.6/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.7/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.8/
drwxrwxrwx  4 root root 4096 May 21  2018 hourly.9/
drwxrwxrwx  3 root root 4096 May  3  2017 Manual_snapshot_2017-11-02/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.0/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.1/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.10/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.11/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.12/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.13/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.14/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.15/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.16/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.17/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.18/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.19/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.2/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.20/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.21/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.22/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.23/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.24/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.25/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.26/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.27/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.28/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.29/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.3/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.4/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.5/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.6/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.7/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.8/
drwxrwxrwx  4 root root 4096 May 21  2018 nightly.9/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.0/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.1/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.10/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.11/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.12/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.13/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.14/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.15/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.16/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.17/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.18/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.19/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.2/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.20/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.21/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.22/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.23/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.3/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.4/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.5/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.6/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.7/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.8/
drwxrwxrwx  4 root root 4096 May 21  2018 weekly.9/

**Comment by No15pecial** [comment]
Wait will Borg still read though the excluded directories  and list them as excluded  The 4 files systems I'm attempting to backup have a lots and lots of snapshots.

**Comment by ThomasWaldmann** [comment]
For a normal "exclude", yes (because there could be another "include" in there).

If you use the more advanced "patterns", there is a "exclude norecurse" that does not do that.

https://borgbackup.readthedocs.io/en/1.1.5/usage/help.html#borg-help-patterns

**Comment by No15pecial** [comment]
Hmm I saw patterns when I fist started looking at borg and I though they would be the better way to go but I couldn't wrap my head around them ... OK time to revisit them.

**Comment by No15pecial** [comment]
OK so 

borg create --pattern=+/a --pattern=-/a/.snapshot --pattern=+/b --pattern=-/b/.snapshot --pattern=+/c/.snapshot.nightl.1 --pattern=+/d/.snapshot/weekly.1 repo::arch /a /b c/.snapshot.nightl.1 /d/.snapshot/weekly.1


or mabe just the following (since /c/.snapshot.nightl.1 and d/.snapshot/weekly.1 are direct and  only contain a single snapshot where /A and /B haev 50 or 100 snapshots) 

borg create --pattern=+/a --pattern=-/a/.snapshot --pattern=+/b --pattern=-/b/.snapshot  repo::arch /a /b c/.snapshot.nightl.1 /d/.snapshot/weekly.1

If I understand the file stuff corectly my script could echo the following to a file /tmp/dailybk.txt

   # "sh:" pattern style is the default, so the following line is not needed:
   P sh
   #Exclude all snapshots in /a
   -/a/.snapshot
   #Exclude all Snalshots in /#
   -/b/.snapshot
   +/a
   +/b
   #Only Backup c/.snapshot.nightl.1 
   +c/.snapshot.nightl.1 
   #Only Backup 
   +/d/.snapshot/weekly.1

then issue the command 
   borg create --patterns-from /tmp/dailybk.txt repo::arch

**Comment by ThomasWaldmann** [comment]
Guess you should also specify the recursion roots (and you do not need to give them as includes).

**Comment by ThomasWaldmann** [comment]
Also, you wanted "exclude norecurse", so it is `!`, not `-`.

**Comment by No15pecial** [comment]
I don't know what this means "Guess you should also specify the recursion roots (and you do not need to give them as includes)."

- One of the problems when you know something so well it seems trivial or like it;s  common knowledge but its not .

**Comment by ThomasWaldmann** [comment]
You usually give the starting path(s) [== recursion roots] as commandline arguments.
You can also give them in the patterns file, iirc it is `R /some/path`.

Of course, when you tell borg to start there / backup them, that is enough, you do not additionally need to "include" them (via `+`).

**Comment by No15pecial** [comment]
Not sure that helped sounds lile your are saying

No need for this 
borg create --pattern=+/a --pattern=-/a/.snapshot --pattern=+/b --pattern=-/b/.snapshot --pattern=+/c/.snapshot.nightl.1 --pattern=+/d/.snapshot/weekly.1 repo::arch /a /b c/.snapshot.nightl.1 /d/.snapshot/weekly.1

Do this instead 

borg create --pattern=-/a/.snapshot --pattern=-/b/.snapshot  repo::arch /a  /b  c/.snapshot.nightl.1  /d/.snapshot/weekly.1

**Comment by ThomasWaldmann** [comment]
Yes. And I would use a pattern file as it is much better readable than that.

**Comment by No15pecial** [comment]
I'll have to modify my script that searches for which snapshot to use and make it write to a file and then use that. 
Thanks for all your assistance - I should be trying the new script tonight

**Comment by No15pecial** [comment]
I guess I'm really dense 

I created a file "TestPAT" containing 
  P sh 
  ! /a/.snapshot
 + /a 

Then I issue the command  and it errors out

root@BorgBK01:~# borg create --list --show-rc --dry-run --patterns-from TestPAT repo::arch
usage: borg [-V] [-h] [--critical] [--error] [--warning] [--info] [--debug]
            [--debug-topic TOPIC] [-p] [--log-json] [--lock-wait SECONDS]
            [--show-version] [--show-rc] [--umask M] [--remote-path PATH]
            [--remote-ratelimit RATE] [--consider-part-files]
            [--debug-profile FILE]
            <command> ...
borg: error: Need at least one PATH argument.

I'm guessing that I need to specify the include in the pattern file and on the command line - Is that correct ?

**Comment by No15pecial** [comment]
Took a bit to test (lots of files and snapshots) but I got it 

Created a file "TestPAT" containing
   P sh
   !  /a/.snapshot

Issue the following command 

borg create --list --show-rc  --patterns-from TestPAT repo::arch /a

I redirected to  >TestPAT.LOG 2>&1  and Tail -f TestPat.LOG so I could see what was happening 

Thanks
